File: rust/varar-runner/src/render.rs
//! Pure human-readable rendering of a step failure, anchored to the `.md`.
//! Port of `var_runner.render.render_failure`; reuses the core diff payloads.

use var_core::error::StepFailure;

pub fn render_failure(failure: &StepFailure, _source: &str, path: &str) -> String {
    let error = &failure.error;
    if let Some(cells) = error.as_cell_mismatch() {
        let mut lines = vec![format!("Cell mismatch in {path}:")];
        let failing: Vec<_> = cells.iter().filter(|c| !c.ok).collect();
        if failing.is_empty() {
            lines.push("  (no failing cells)".to_string());
        }
        for cell in failing {
            lines.push(format!(
                "  line {} | column '{}' — expected: {:?}, actual: {:?}",
                cell.span.start_line, cell.column, cell.expected, cell.actual
            ));
        }
        return lines.join("\n");
    }
    if let Some(diff) = error.as_doc_string_mismatch() {
        return format!(
            "Doc string mismatch at line {}:\n  expected: {:?}\n  actual:   {:?}",
            diff.span.start_line, diff.expected, diff.actual
        );
    }
    error.message()
}

File: typescript/packages/runner/src/render.ts
import { isCellMismatchError, isDocStringMismatchError, ReturnShapeError } from '@varar/core'

/**
 * Render a step failure as a human-readable string, anchored to the source `.md`
 * location when the error carries structured diff information.
 *
 * Structured errors (`CellMismatchError`, `DocStringMismatchError`,
 * `ReturnShapeError`) produce expected/actual output with the `.md` line number.
 * All other throws fall back to `error.stack` or `String(error)`.
 */
export function renderFailure(error: unknown, _source: string, path: string): string {
  if (isCellMismatchError(error)) {
    const failingCells = error.cells.filter((c) => !c.ok)
    if (failingCells.length === 0) return error.stack ?? error.message
    const lines = failingCells.map((c) => {
      const line = c.span.startLine
      return `  ${path}:${line} col "${c.column}": expected ${JSON.stringify(c.expected)} but was ${JSON.stringify(c.actual)}`
    })
    return `CellMismatchError\n${lines.join('\n')}`
  }

  if (isDocStringMismatchError(error)) {
    const line = error.diff.span.startLine
    return [
      `DocStringMismatchError at ${path}:${line}`,
      `  expected: ${JSON.stringify(error.diff.expected)}`,
      `  actual:   ${JSON.stringify(error.diff.actual)}`,
    ].join('\n')
  }

  if (error instanceof ReturnShapeError) {
    return `ReturnShapeError: ${error.message}`
  }

  if (error instanceof Error && typeof error.stack === 'string') return error.stack
  if (error instanceof Error) return error.message
  return String(error)
}
